manager-rs/dauth-service/src/tasks/replace_key_shares.rs

```diff
@@ -0,0 +1,68 @@
+use std::sync::Arc;
+
+use crate::data::context::DauthContext;
+use crate::data::error::DauthError;
+use crate::database;
+use crate::database::tasks::replace_key_shares::ReplaceKeyShareTask;
+use crate::rpc::clients;
+
+/// Runs the replace key shares task.
+/// If any replacements are queued up, waits for 10 seconds and then calls
+/// the key share replacement rpc.
+pub async fn run_task(context: Arc<DauthContext>) -> Result<(), DauthError> {
+    let mut transaction = context.local_context.database_pool.begin().await.unwrap();
+    let replaces = database::tasks::replace_key_shares::get(&mut transaction).await?;
+    transaction.commit().await.unwrap();
+
+    if replaces.is_empty() {
+        tracing::info!("Nothing to do for replace key share task");
+    } else {
+        tracing::info!("Found {} replace key share(s) pending", replaces.len());
+
+        let mut tasks = Vec::new();
+
+        for replace in replaces {
+            tasks.push(tokio::spawn(replace_key_share(context.clone(), replace)));
+        }
+
+        for task in tasks {
+            match task.await {
+                Ok(task_res) => match task_res {
+                    Ok(replace) => {
+                        let mut transaction = context.local_context.database_pool.begin().await?;
+                        database::tasks::replace_key_shares::remove(
+                            &mut transaction,
+                            &replace.backup_network_id,
+                            &replace.xres_star_hash,
+                        )
+                        .await?;
+                        transaction.commit().await?;
+                    }
+                    Err(e) => {
+                        tracing::warn!("Failed to execute replace task: {}", e)
+                    }
+                },
+                Err(je) => {
+                    tracing::warn!("Error while joining: {}", je)
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+async fn replace_key_share(
+    context: Arc<DauthContext>,
+    replace: ReplaceKeyShareTask,
+) -> Result<ReplaceKeyShareTask, DauthError> {
+    let mut replace_delay = tokio::time::interval(context.tasks_context.replace_key_share_delay);
+    replace_delay.tick().await; // first tick does nothing
+    replace_delay.tick().await;
+
+    let (address, _) =
+        clients::directory::lookup_network(context.clone(), &replace.backup_network_id).await?;
+
+    clients::backup_network::replace_key_share(context, &replace, &address).await?;
```

Comment:
Yes, it's handled at a higher level exactly how you think. The db is used as a queue, and tasks are only removed on success (a later goal is check specific errors or conditions other than success to remove them).

The delay tick is run every time, even though it doesn't have to be. I did this to avoid having to save state (i.e. either when the task was queued or if the wait already happened).

Since this is queued from the server side, it didn't seem appropriate to wait before adding to the queue. That said, I could spawn a thread that waits instead, but that would stop me from detecting errors with the db on the server side.

What are your thoughts on this?

Reply:
I think always delaying here is okay to not need a special case (since this doesn't need to be highly responsive), but these internal delays will interact with the top-level task loop delay given the current always-await design. With these delays now in the event that the network cannot be reached we'll always await at the top level for max(replace_key_share_delay + tonic_timeout, tasks_context.interval).

Reply:
That's a really good point. If we want to avoid this, I could add some in-memory state. Something like a "first seen" time

Reply:
I don't think it's essential again given that this process doesn't have any strict timing requirements besides "eventually" : ) If in testing we realize we're missing an edge case where this needs to complete quickly we can revisit it.